### PR TITLE
Added support for always-auth and _auth. Both for root and scopes #16528

### PR DESCRIPTION
**Pull request by jozemlakar**
I had added the loading of auth config when always-auth and _auth are set.
I had added the sending of authorization header in the mentioned case. 
I have tested this against artifactory, when set globally. Someone else might want to test it against other registries or when scopes are used.

### CODE REVIEW

File: lib/config/pacote.js
Comment (snirad) [inline lib/config/pacote.js]: this regex assume we use 
//registry.npmjs... type of syantax.

could you please change it like so ?

/((.*):)?(_authToken|username|_password|password|email|always-auth|_auth)$/

**Review by snirad — COMMENTED** [review]
- sorry for the confusion , now i see that it works even without adding this regex, but maybe it is still needed, im not sure.

File: node_modules/pacote/lib/fetchers/registry/fetch.js
Comment (zkat) [inline node_modules/pacote/lib/fetchers/registry/fetch.js]: These changes should be files as a separate PR against `pacote`.

File: lib/config/pacote.js
Comment (zkat) [inline lib/config/pacote.js]: I'd rather not coerce to boolean here. Let `if` do its usual work?

File: lib/config/pacote.js
Comment (zkat) [inline lib/config/pacote.js]: Please make sure the code passes `standard` linting.

File: lib/config/pacote.js
Comment (zkat) [inline lib/config/pacote.js]: This can't just be for `_auth`. All of the other auth parameters listed below should be made available as toplevel settings.

**Review by zkat — CHANGES_REQUESTED** [review]
So far so good! A few things to improve first 👍

File: lib/config/pacote.js
Comment (snirad) [inline lib/config/pacote.js]: how about adding this line just below, and making authMatch ; let instead of const.
that should hold for all cases, and you can remove that if condition you added just for _auth

`
  npm.config.keys.forEach(function (k) {
    let authMatch = (k[0] === '/' && k.match(/(.*):(_authToken|username|_password|password|email|always-auth|_auth)$/))
   if(!authMatch && /(_authToken|username|_password|password|email|always-auth|_auth)$/.test(k)){
        authMatch =[,registryKey(opts.registry), k]
}
    if (authMatch) {
      const nerfDart = authMatch[1]
      const key = authMatch[2]
      const val = npm.config.get(k)
.........................

`

- also we will need to import registryKey functoin as wel

`function registryKey (registry) {
  const parsed = url.parse(registry)
  const formatted = url.format({
    host: parsed.host,
    pathname: parsed.pathname,
    slashes: parsed.slashes
  })
  return url.resolve(formatted, '.')
}`

what do you think,
@zkat , @jozemlakar

Comment (jozemlakar) [reply lib/config/pacote.js]: sounds good. I am currently working on tests on a change in https://github.com/zkat/pacote and will then do config. Whatever the final solution, it should be the same for scoped and global settings, so my first one is no good.

File: node_modules/pacote/lib/fetchers/registry/fetch.js
Comment (jozemlakar) [reply node_modules/pacote/lib/fetchers/registry/fetch.js]: Added https://github.com/zkat/pacote/pull/96
This has to be verified and then tagged so we can upgrade dependency in npm. Then I will try and fix the config.

File: lib/config/pacote.js
Comment (jozemlakar) [reply lib/config/pacote.js]: So, looked at this issue and here is where I stand:
- PR created for pacote. I think that should clear. Once that lands we can update the dependency in npm
- There is only one general test for config/pacote in npm, so the first thing I need to do is create a test for config parsing
- Fix config parsing so we get all options (including _auth) both on global and scoped.

**Comment by jozemlakar** [comment]
I think this is now complete. It adds support for all auth options, uses options in the same manner when scoped and global.

**Review by zkat — APPROVED** [review]

**Comment by zkat** [comment]
Awesome work! Yay tests! I'll merge this once the test suite passes. Thanks a bunch 🎉

**Comment by zkat** [comment]
@jozemlakar the tests seem to be failing right now :< https://travis-ci.org/npm/npm/builds/240593681?utm_source=github_status&utm_medium=notification

**Comment by jozemlakar** [comment]
@zkat seems that registry configured in test is no longer used but rather goes to default public registry. Will look into it.

**Comment by jozemlakar** [comment]
@zkat Build was broken with https://github.com/npm/npm/pull/17076
when merged without running tests first https://github.com/npm/npm/pull/16879/commits/c8528297ada8a542385efbefa6ca0d776bbb6342

Without that change the config is
```
{ type: 'remote',
  registry: undefined,
  where: undefined,
  raw: '@scoped/underscore@http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  name: '@scoped/underscore',
  escapedName: '@scoped%2funderscore',
  scope: '@scoped',
  rawSpec: 'http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  saveSpec: 'http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  fetchSpec: 'http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  gitRange: undefined,
  gitCommittish: undefined,
  hosted: undefined }
```

after change the config is
```
{ type: 'version',
  registry: true,
  where: undefined,
  raw: '@scoped/underscore@1.3.1',
  name: '@scoped/underscore',
  escapedName: '@scoped%2funderscore',
  scope: '@scoped',
  rawSpec: '1.3.1',
  saveSpec: null,
  fetchSpec: '1.3.1',
  gitRange: undefined,
  gitCommittish: undefined,
  hosted: undefined }
```

As a result, npm is sending the request to public npm registry instead of the mock.

I have no idea why this was changed or how this passed existing tests. Please have a look at it.

**Comment by jozemlakar** [comment]
@zkat if you need me to check something or try something out, let me know. My knowledge of shrinkwrap is next to zero, so I really can't handle the changes you made on my own.

**Comment by jozemlakar** [comment]
@zkat in case I was not understood. Commit https://github.com/npm/npm/commit/c8528297ada8a542385efbefa6ca0d776bbb6342 breaks existing tests i.e. shrinkwrap-scoped-auth.js
When you fix that I can fix other tests I created based on that one.

**Comment by jozemlakar** [comment]
@snirad you are absolutely correct! I remove the lines, thank you.

**Comment by jozemlakar** [comment]
@zkat this is now ready. Once i pulled all the changes from release-next and synced my tests with the ones you changed tests succeed.

**Comment by snirad** [comment]
@zkat  hi kat :) 

any chance we get to see it on 5.0.4 ?

**Comment by zkat** [comment]
This has been merged into `release-next` and is available in `npmc@5.0.4-canary.4`!

**Pull request closed without merging**
